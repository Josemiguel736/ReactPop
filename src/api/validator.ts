import { z } from 'zod';
import { AdvertType } from '../pages/adverts/types';
import { ApiClientError } from './error';
import { Login, Me } from '../pages/auth/types';

const advertsSchema = z.object({
	id: z.string(),
	name: z.string(),
	createdAt: z.string(),
	sale: z.boolean(),
	price: z.number(),
	tags: z.array(z.string()),
	photo: z.string().nullable(),
});

const tagsSchema = z.array(z.string());

const loginSchema = z.object({
	accessToken: z.string(),
});

const meSchema = z.object({
	id: z.string(),
	createdAt: z.string(),
	email: z.string(),
	username: z.string(),
	name: z.string(),
});

export const validateAdverts = (data: AdvertType[]) => {
	try {
		data.forEach((advert) => {
			advertsSchema.parse(advert);
		});
		return true;
	} catch (error) {
		throw new ApiClientError('Error validating adverts', 'SERVER_ERROR');
	}
};

export const validateAdvert = (data: AdvertType) => {
	try {
		advertsSchema.parse(data);
		return true;
	} catch (error) {
		throw new ApiClientError('Error validating advert', 'SERVER_ERROR');
	}
};

export const validateTags = (data: string[]) => {
	try {
		tagsSchema.parse(data);
		return true;
	} catch (error) {
		throw new ApiClientError('Error validating tags', 'SERVER_ERROR');
	}
};

export const validateLogin = (data: Login) => {
	try {
		loginSchema.parse(data);
		return true;
	} catch (error) {
		throw new ApiClientError('Error validating login', 'SERVER_ERROR');
	}
};

export const validateMe = (data: Me) => {
	try {
		meSchema.parse(data);
		return true;
	} catch (error) {
		throw new ApiClientError('Error validating Me', 'SERVER_ERROR');
	}
};
