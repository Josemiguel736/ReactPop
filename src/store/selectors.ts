import { RootState } from ".";
import { AdvertType } from "../pages/adverts/types";

const defaultAdvertData: AdvertType[] = []
const defaultTagsData : string[] =[]

export const getIsLogged = (state:RootState) => state.auth

export const getUi = (state:RootState)=>state.ui

export const getAdverts = (state:RootState) => state.adverts.data || defaultAdvertData

export const getTags = (state:RootState) => state.tags.data ?? defaultTagsData

export const getTagsError = (state:RootState) => state.tags.error?? null

export const getAdvert = (advertId?:string) => (state:RootState) =>
    state.adverts.data?.find((advert) => advert.id === advertId)