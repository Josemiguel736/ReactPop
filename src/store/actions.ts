import { AppThunk } from '.';
import { isApiClientError } from '../api/client';
import { AdvertType } from '../pages/adverts/types';
import { Credentials } from '../pages/auth/types';
import { getAdvert, removeAdvert } from './selectors';

// ------------------------------- Login -------------------------------------------------

type AuthLoginPending = {
	type: 'auth/login/pending';
};

type AuthLoginFulfilled = {
	type: 'auth/login/fulfilled';
};

type AuthLogout = {
	type: 'auth/logout';
};

type AuthLoginRejected = {
	type: 'auth/login/rejected';
	payload: Error;
};

// ----------------------- Adverts -------------------------------------------

type AdvertsLoadedPending = {
	type: 'adverts/loaded/pending';
};

type AdvertsLoadedFulfilled = {
	type: 'adverts/loaded/fulfilled';
	payload: { data: AdvertType[]; loaded: boolean };
};

type AdvertsLoadedRejected = {
	type: 'adverts/loaded/rejected';
	payload: Error;
};

// --------------------------- Advert ---------------------------------------------

type AdvertCreatedFulfilled = {
	type: 'advert/created/fulfilled';
	payload: AdvertType;
};

type AdvertCreatedRejected = {
	type: 'advert/created/rejected';
	payload: Error
};

type AdvertDeletedFulfilled = {
	type: 'advert/deleted/fulfilled';
	payload: AdvertType[]
};

type AdvertDeletedRejected = {
	type: 'advert/deleted/rejected';
	payload: Error
};

type AdvertLoadedFulfilled = {
	type: 'advert/loaded/fulfilled';
	payload: AdvertType;
};

type AdvertLoadedRejected = {
	type: 'advert/loaded/rejected';
	payload: Error
};

type AdvertLoadedPending = {
	type: 'advert/loaded/pending';
};

// --------------------------- Tags --------------------------------------------

type TagsLoadedFulfilled = {
	type: 'tags/loaded/fulfilled';
	payload:  string[] ;
};

type TagsLoadedRejected = {
	type: 'tags/loaded/rejected';
	payload: Error;
};

type TagsLoadedPending = {
	type: 'tags/loaded/pending';
};

// -------------------------- UI ----------------------------------------------
type UiResetError = {
	type: 'ui/reset-error';
};

// ----------------------- Login ----------------------------------------------

export const authLoginPending = (): AuthLoginPending => ({
	type: 'auth/login/pending',
});

export const authLoginFulfilled = (): AuthLoginFulfilled => ({
	type: 'auth/login/fulfilled',
});

export const AuthLogout = (): AuthLogout => ({
	type: 'auth/logout',
});

export const authLoginRejected = (error: Error): AuthLoginRejected => ({
	type: 'auth/login/rejected',
	payload: error,
});

export function authLogin(
	credentials: Credentials,
	remember: boolean,
): AppThunk<Promise<void>> {
	return async function (dispatch, _getState, { api, router }) {
		dispatch(authLoginPending());
		try {
			await api.auth.login(credentials, remember);
			dispatch(authLoginFulfilled());
			const to = router.state.location.state?.from ?? '/';
			router.navigate(to, { replace: true });
		} catch (error) {
			if (isApiClientError(error)) {
				dispatch(authLoginRejected(error));
			} else {
				console.log('ERROR ', error); //TODO
			}
		}
	};
}

// -------------------------- Adverts -----------------------------------------

export const advertsLoadedFulfilled = (
	adverts: AdvertType[],
	loaded?: boolean,
): AdvertsLoadedFulfilled => ({
	type: 'adverts/loaded/fulfilled',
	payload: { data: adverts, loaded: !!loaded },
});

export function advertsLoaded(): AppThunk<Promise<void>> {
	return async function (dispatch, getState, { api }) {
		const state = getState();
		if (state.adverts.loaded) {
			return;
		}
		dispatch(advertsLoadedPending());
		try {
			const adverts = await api.adverts.getLastestAdverts();
			dispatch(advertsLoadedFulfilled(adverts, true));
		} catch (error) {
			if (isApiClientError(error)) {
				dispatch(advertsLoadedRejected(error));
			} else {
				console.log(error);
			}
		}
	};
}

export const advertsLoadedRejected = (error: Error): AdvertsLoadedRejected => ({
	type: 'adverts/loaded/rejected',
	payload: error,
});

export const advertsLoadedPending = (): AdvertsLoadedPending => ({
	type: 'adverts/loaded/pending',
});

// --------------------------- Advert ----------------------------------

export const advertLoadedPending = (): AdvertLoadedPending => ({
	type: 'advert/loaded/pending',
});

export const advertCreatedFulfilled = (advert:AdvertType):AdvertCreatedFulfilled =>({
	type: "advert/created/fulfilled",
	payload:advert
})

 export const advertDeletedFulfilled = (adverts:AdvertType[]):AdvertDeletedFulfilled =>({
 	type: "advert/deleted/fulfilled",
	payload:adverts
 })

export const advertDeletedRejected = (error:Error):AdvertDeletedRejected =>({
	type: "advert/deleted/rejected",
	payload:error
})

export const advertCreatedRejected = (error:Error):AdvertCreatedRejected =>({
	type: "advert/created/rejected",
	payload:error
})

export function advertDeleted(advertId:string):AppThunk<Promise<void>>{
	return async function(dispatch,getState,{api,router}){
		try {
			dispatch(advertLoadedPending())
			const state = getState()
			
			await api.adverts.deleteAdvert(advertId)
			const adverts = removeAdvert(advertId)(state)
			
			dispatch(advertDeletedFulfilled(adverts ?? []))
		
		router.navigate("/")
			
		} catch (error) {
			if(isApiClientError(error)){
				dispatch(advertDeletedRejected(error))
			}
			console.log(error)
			
		}
	}

}

export function advertCreated(advertContent:FormData):AppThunk<Promise<AdvertType>>{
	return async function(dispatch,_getState,{api,router}) {
		try {
			dispatch(advertLoadedPending())
			const createdAdvert = await api.adverts.createAdvert(advertContent)
			const advert = await api.adverts.getAdvert(createdAdvert.id)
			dispatch(advertCreatedFulfilled(advert))
			await router.navigate(`/adverts/${advert.id}`)
			return advert
		} catch (error) {
			if (isApiClientError(error)){
				dispatch(advertCreatedRejected(error))
			
			}
			throw error}
	}}

export const advertLoadedRejected = (error:Error):AdvertLoadedRejected =>({
	type:'advert/loaded/rejected',
	payload: error
})

export function advertLoaded(advertId:string): AppThunk<Promise<void>>{
	return async function(dispatch, getState, {api,router}){
		const state = getState()
		if(getAdvert(advertId)(state)){
			return
		}
		try {
			dispatch(advertLoadedPending())
			const advert = await api.adverts.getAdvert(advertId)
			dispatch(advertsLoadedFulfilled([advert]))
		} catch (error) {
			if(isApiClientError(error)){
				dispatch(advertLoadedRejected(error))
			}
			router.navigate("/404")
			
		}
	}
}
		

// -------------------------------- Tags ------------------------------------

export const tagsLoadedFulfilled = (
	tags: string[]
): TagsLoadedFulfilled => ({
	type: 'tags/loaded/fulfilled',
	payload:  tags
});

export const tagsLoadedPending = (): TagsLoadedPending => ({
	type: 'tags/loaded/pending',
});

export const tagsLoadedRejected = (error: Error): TagsLoadedRejected => ({
	type: 'tags/loaded/rejected',
	payload: error,
});

export function tagsLoaded(): AppThunk<Promise<void>> {
	return async function (dispatch, getState, { api }) {
		const state = getState();
		if (state.tags.loaded) {
			return;
		}
		try {
			dispatch(tagsLoadedPending());
			const tags = await api.adverts.getTags();
			dispatch(tagsLoadedFulfilled(tags));
		} catch (error) {
			if (isApiClientError(error)) {
				dispatch(tagsLoadedRejected(error));
			} else {
				console.log(error);
			}
		}
	};
}

// ------------------------------ UI ---------------------------------------

export const UiResetError = (): UiResetError => ({
	type: 'ui/reset-error',
});

export type Actions =
	| AuthLoginFulfilled
	| AuthLogout
	| AuthLoginPending
	| AuthLoginRejected
	| AdvertsLoadedFulfilled
	| AdvertsLoadedRejected
	| AdvertCreatedFulfilled
	| UiResetError
	| AdvertsLoadedPending
	| TagsLoadedFulfilled
	| TagsLoadedPending
	| TagsLoadedRejected
	| AdvertCreatedRejected
	| AdvertCreatedFulfilled
	| AdvertLoadedPending
	| AdvertLoadedRejected
	| AdvertLoadedFulfilled
	| AdvertDeletedFulfilled
	| AdvertDeletedRejected