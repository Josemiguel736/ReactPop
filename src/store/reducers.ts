import { AdvertType } from '../pages/adverts/types';
import { Actions } from './actions';

export type State = {
	auth: boolean;
	adverts: { data: AdvertType[] | null; loaded: boolean };
	tags: { data: string[] | null; loaded: boolean };
	ui: {
		pending: boolean;
		error: Error | null;
		tagsError: Error | null;
	};
};

const defaultState: State = {
	auth: false,
	adverts: {
		data: null,
		loaded: false,
	},
	tags: { data: null, loaded: false },
	ui: {
		pending: false,
		error: null,
		tagsError: null,
	},
};

export function auth(
	state = defaultState.auth,
	action: Actions,
): State['auth'] {
	switch (action.type) {
		case 'auth/login/fulfilled':
			return true;
		case 'auth/logout':
			return false;
		default:
			return state;
	}
}

export function ui(state = defaultState.ui, action: Actions): State['ui'] {
	switch (action.type) {
		case 'ui/reset-error':
			return { ...state, error: null };

		case 'auth/login/pending':
			return { error: null, pending: true, tagsError: null };
		case 'adverts/loaded/pending':
			return { error: null, pending: true, tagsError: null };
		case 'advert/loaded/pending':
			return { error: null, pending: true, tagsError: null };
		case 'tags/loaded/pending':
			return { error: null, pending: true, tagsError: null };

		case 'auth/login/rejected':
			return { ...state, pending: false, error: action.payload };
		case 'adverts/loaded/rejected':
			return { ...state, pending: false, error: action.payload };
		case 'advert/deleted/rejected':
			return { ...state, pending: false, error: action.payload };
		case 'advert/loaded/rejected':
			return { ...state, pending: false, error: action.payload };
		case 'advert/created/rejected':
			return { ...state, pending: false, error: action.payload };
		case 'tags/loaded/rejected':
			return { ...state, pending: false, tagsError: action.payload };

		case 'auth/login/fulfilled':
			return { ...state, pending: false, error: null };
		case 'tags/loaded/fulfilled':
			return { ...state, pending: false, tagsError: null };
		case 'adverts/loaded/fulfilled':
			return { ...state, pending: false, error: null };
		case 'advert/created/fulfilled':
			return { ...state, pending: false, error: null };
		case 'advert/deleted/fulfilled':
			return { ...state, pending: false, error: null };
		default:
			return state;
	}
}

export function adverts(
	state = defaultState.adverts,
	action: Actions,
): State['adverts'] {
	switch (action.type) {
		case 'adverts/loaded/fulfilled':
			return action.payload;
		case 'advert/created/fulfilled':
			return { ...state, data: [...(state.data ?? []), action.payload] };
		case 'advert/deleted/fulfilled':
			return { ...state, data: action.payload };
		default:
			return state;
	}
}

export function tags(
	state = defaultState.tags,
	action: Actions,
): State['tags'] {
	switch (action.type) {
		case 'tags/loaded/fulfilled':
			return {
				...state,
				data: Array.isArray(action.payload) ? action.payload : [],
				loaded: true,
			};
		case 'tags/loaded/rejected':
			return { ...state, loaded: false };
		default:
			return state;
	}
}
