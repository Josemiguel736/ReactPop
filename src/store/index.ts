import {applyMiddleware, combineReducers, createStore} from "redux"
import * as reducers from "./reducers"
import {State} from "./reducers"
import { createBrowserRouter } from "react-router-dom"
import * as auth from "../pages/auth/service"
import * as adverts from "../pages/adverts/service"
import { composeWithDevTools } from "@redux-devtools/extension"
import * as thunk from "redux-thunk"
import { Actions } from "./actions"
import { useDispatch, useSelector } from "react-redux"

type Router = ReturnType<typeof createBrowserRouter>

type Api = {
    auth:typeof auth
    adverts:typeof adverts
}

type ExtraArgument = {
    api:Api
    router:Router
}

export default function configureStore(preloadedState:Partial<State>,router:Router){
    const rootReducer = combineReducers(reducers)
    const store = createStore(
        rootReducer,
        preloadedState as never,
        composeWithDevTools(
            applyMiddleware(thunk.withExtraArgument<State,Actions,ExtraArgument>({api:{auth,adverts},router}))
        )
        
    )
    return store;
}

export type AppStore = ReturnType<typeof configureStore>
export type AppGetState = AppStore["getState"]
export type RootState = ReturnType<AppGetState>
export type AppDispatch = AppStore["dispatch"]

export const useAppDispatch = useDispatch.withTypes<AppDispatch>()
export const useAppSelector = useSelector.withTypes<RootState>()

export type AppThunk<ReturnType = void> = thunk.ThunkAction<
  ReturnType,
  RootState,
  ExtraArgument,
  Actions
>;