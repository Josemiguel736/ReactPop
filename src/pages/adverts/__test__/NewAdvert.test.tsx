import {render, screen} from "@testing-library/react"
import { Store } from "redux"
import { State } from "../../../store/reducers"
import { Provider } from "react-redux"
import NewAdvertPage from "../NewAdvert"
import userEvent from "@testing-library/user-event"
import * as actions from "../../../store/actions"
import { ApiClientError } from "../../../api/error"

vi.mock("../../../store/actions")

describe("NewAdvertPage",() => {
    afterEach(()=>{
        vi.clearAllMocks()
    })

const tags:string[] = ["motor","lifestile","work"]

const initialState:State = {
    auth:true,
    adverts:{data:null,loaded:false},
   tags:{data:tags,loaded:false,error:null},
    ui:{error:null,pending:false}    }


const mockStore: Partial<Store<State>> = {
    getState: () => initialState,
    subscribe: vi.fn(),
    dispatch: vi.fn(),
  };

  const formValues = {
    product:"Coche",
    price:"2000",
    img: new File(["Test image"],"img.png",{type:"image/png"})
  }

  const mockFormData:Partial<FormData>  = {
    append: vi.fn()
  }

const renderComponent = (error?:Error|null,tagError?:ApiClientError) =>{
     initialState.ui.error = error ?? null
     initialState.tags.error = tagError ?? null

    return(
    render(<Provider store={mockStore as Store<State>}>
     <NewAdvertPage/></Provider>))}

    test("should render",() => {
        const {container} = renderComponent()
        expect(container).toMatchSnapshot()       
    })

    test("should dispatch new advert action", async () => {
        renderComponent()
        globalThis.FormData = vi.fn(()=>mockFormData as FormData) 

        const nameInput = screen.getByPlaceholderText("Producto")
        await userEvent.type(nameInput,formValues.product)
        expect(nameInput).toHaveValue(formValues.product)

        const priceInput = screen.getByPlaceholderText("Precio")
        await userEvent.type(priceInput,formValues.price)
        expect(priceInput).toHaveValue(Number(formValues.price))

        const selectElement = screen.getByRole("combobox")
        expect(selectElement).toBeInTheDocument()

        const buyOption = screen.getByRole("option",{name:"Compra"})
        expect(buyOption).toBeInTheDocument()

        
        const saleOption = screen.getByRole("option",{name:"Venta"})
        expect(saleOption).toBeInTheDocument()
        
        await userEvent.selectOptions(selectElement,"Compra")

        const lifestileTag = screen.getByText("lifestile")
        await userEvent.click(lifestileTag)

        const motorTag = screen.getByText("motor")
        await userEvent.click(motorTag)

        const workTag = screen.getByText("work")
        await userEvent.click(workTag)
        await userEvent.click(workTag)

        const imgInput = screen.getByLabelText("Sube una foto") as HTMLInputElement
        await userEvent.upload(imgInput,formValues.img)
        expect(imgInput.files).toHaveLength(1)
        expect(imgInput.files?.[0]).toBe(formValues.img)

        const submitButton = screen.getByRole("button")
        await userEvent.click(submitButton)

        expect(actions.advertCreated).toBeCalled()
        expect(mockFormData.append).toHaveBeenCalledWith('name', formValues.product);
        expect(mockFormData.append).toHaveBeenCalledWith('price', formValues.price);
        expect(mockFormData.append).toHaveBeenCalledWith('tags', "lifestile");
        expect(mockFormData.append).toHaveBeenCalledWith('tags', "motor");
        expect(mockFormData.append).toHaveBeenCalledWith('sale', "false");
    })

        test("Test to tags render error", async ()=>{
            const {container} = renderComponent(null,new ApiClientError("Not_found","NOT_FOUND"))        
            const alert = screen.getByRole("alert")
    
            expect(alert).toHaveTextContent("Ha surgido un error al cargar los tags por favor intentelo más tarde")
            expect(container).toMatchSnapshot()     
    
        })

        test("Test to render error", async ()=>{

            const {container} = renderComponent(new ApiClientError("Not_found","NOT_FOUND"))        
            const alert = screen.getByRole("alert")
    
            expect(alert).toHaveTextContent("Ha ocurrido un problema al crear el producto porfavor intentalo más tarde")
            expect(container).toMatchSnapshot()     
    
        })

        test("Test to minText error", async ()=>{
            formValues.product = "aa"
            renderComponent()        
            const nameInput = screen.getByPlaceholderText("Producto")
            await userEvent.type(nameInput,formValues.product)
            expect(nameInput).toHaveValue(formValues.product)

            const alert = screen.getByRole("alert")    
            expect(alert).toHaveTextContent("El producto debe de tener al menos 3 letras")
            const submitButton = screen.getByRole("button")
            await userEvent.click(submitButton)

            expect(actions.advertCreated).not.toBeCalled()
        })

        test("Test to number invalid error", async ()=>{
            formValues.price = "-50"
            renderComponent()        
            const priceInput = screen.getByPlaceholderText("Precio")
            await userEvent.type(priceInput,formValues.price)
            expect(priceInput).toHaveValue(Number(formValues.price))

            const alert = screen.getByRole("alert")    
            expect(alert).toHaveTextContent("El precio no puede ser negativo")
            const submitButton = screen.getByRole("button")
            await userEvent.click(submitButton)

            expect(actions.advertCreated).not.toBeCalled()
        })




})