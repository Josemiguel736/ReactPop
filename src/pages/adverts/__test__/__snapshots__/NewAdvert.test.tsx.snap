// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewAdvertPage > Test to render error 1`] = `
<div>
  <div
    class="h-1/3 mt-10 flex items-center justify-center"
  >
    <form
      class="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
    >
      <h1
        class="text-2xl mt-2.5 mb-1"
      >
        Crear anuncio
      </h1>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="name"
            placeholder="Producto"
            required=""
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="price"
            placeholder="Precio"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <select
        class="cursor-pointer"
        required=""
      >
        <option
          class="cursor-pointer bg-gray-500"
          value="compra"
        >
          Compra
        </option>
        <option
          class="cursor-pointer bg-gray-500"
          value="venta"
        >
          Venta
        </option>
      </select>
      <h3
        class="text-xl mb-4"
      >
        Selecciona tus categorías:
      </h3>
      <div
        class="flex flex-col gap-2 "
      >
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="motor"
          />
          motor
        </label>
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="lifestile"
          />
          lifestile
        </label>
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="work"
          />
          work
        </label>
      </div>
      <div
        class="mt-4 p-3 rounded-lg "
      >
        <strong>
          Seleccionados:
        </strong>
         
        Ninguno
      </div>
      <label
        for="file-upload"
      >
        Sube una foto
      </label>
      <input
        class="bg-sky-900 p-2.5 w-full sm:w-100 cursor-pointer "
        id="file-upload"
        type="file"
      />
      <button
        class="sc-FEMpB dyFOCA mt-4 mb-2.5"
        type="submit"
      >
        Publicar
      </button>
      <span
        class="text-red-500 bg-sky-950/50 pl-3.5 pr-3.5 "
        role="alert"
      >
        Ha ocurrido un problema al crear el producto porfavor intentalo más tarde
      </span>
    </form>
  </div>
</div>
`;

exports[`NewAdvertPage > Test to tags render error 1`] = `
<div>
  <div
    class="h-1/3 mt-10 flex items-center justify-center"
  >
    <form
      class="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
    >
      <h1
        class="text-2xl mt-2.5 mb-1"
      >
        Crear anuncio
      </h1>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="name"
            placeholder="Producto"
            required=""
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="price"
            placeholder="Precio"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <select
        class="cursor-pointer"
        required=""
      >
        <option
          class="cursor-pointer bg-gray-500"
          value="compra"
        >
          Compra
        </option>
        <option
          class="cursor-pointer bg-gray-500"
          value="venta"
        >
          Venta
        </option>
      </select>
      <h3
        class="text-xl mb-4"
      >
        Selecciona tus categorías:
      </h3>
      <span
        class="text-red-500 bg-sky-950/50 pl-3.5 pr-3.5 "
        role="alert"
      >
        Ha surgido un error al cargar los tags por favor intentelo más tarde
      </span>
      <div
        class="mt-4 p-3 rounded-lg "
      >
        <strong>
          Seleccionados:
        </strong>
         
        Ninguno
      </div>
      <label
        for="file-upload"
      >
        Sube una foto
      </label>
      <input
        class="bg-sky-900 p-2.5 w-full sm:w-100 cursor-pointer "
        id="file-upload"
        type="file"
      />
      <button
        class="sc-FEMpB dyFOCA mt-4 mb-2.5"
        type="submit"
      >
        Publicar
      </button>
    </form>
  </div>
</div>
`;

exports[`NewAdvertPage > Test to tags render error 2`] = `
<div>
  <div
    class="h-1/3 mt-10 flex items-center justify-center"
  >
    <form
      class="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
    >
      <h1
        class="text-2xl mt-2.5 mb-1"
      >
        Crear anuncio
      </h1>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="name"
            placeholder="Producto"
            required=""
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="price"
            placeholder="Precio"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <select
        class="cursor-pointer"
        required=""
      >
        <option
          class="cursor-pointer bg-gray-500"
          value="compra"
        >
          Compra
        </option>
        <option
          class="cursor-pointer bg-gray-500"
          value="venta"
        >
          Venta
        </option>
      </select>
      <h3
        class="text-xl mb-4"
      >
        Selecciona tus categorías:
      </h3>
      <div
        class="flex flex-col gap-2 "
      >
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="motor"
          />
          motor
        </label>
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="lifestile"
          />
          lifestile
        </label>
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="work"
          />
          work
        </label>
      </div>
      <div
        class="mt-4 p-3 rounded-lg "
      >
        <strong>
          Seleccionados:
        </strong>
         
        Ninguno
      </div>
      <label
        for="file-upload"
      >
        Sube una foto
      </label>
      <input
        class="bg-sky-900 p-2.5 w-full sm:w-100 cursor-pointer "
        id="file-upload"
        type="file"
      />
      <button
        class="sc-FEMpB dyFOCA mt-4 mb-2.5"
        type="submit"
      >
        Publicar
      </button>
      <span
        class="text-red-500 bg-sky-950/50 pl-3.5 pr-3.5 "
        role="alert"
      >
        Ha ocurrido un problema al crear el producto porfavor intentalo más tarde
      </span>
    </form>
  </div>
</div>
`;

exports[`NewAdvertPage > should render 1`] = `
<div>
  <div
    class="h-1/3 mt-10 flex items-center justify-center"
  >
    <form
      class="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
    >
      <h1
        class="text-2xl mt-2.5 mb-1"
      >
        Crear anuncio
      </h1>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="name"
            placeholder="Producto"
            required=""
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          <input
            class="mt-4 border-2 rounded-lg "
            name="price"
            placeholder="Precio"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <select
        class="cursor-pointer"
        required=""
      >
        <option
          class="cursor-pointer bg-gray-500"
          value="compra"
        >
          Compra
        </option>
        <option
          class="cursor-pointer bg-gray-500"
          value="venta"
        >
          Venta
        </option>
      </select>
      <h3
        class="text-xl mb-4"
      >
        Selecciona tus categorías:
      </h3>
      <div
        class="flex flex-col gap-2 "
      >
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="motor"
          />
          motor
        </label>
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="lifestile"
          />
          lifestile
        </label>
        <label
          class="flex items-center gap-2 cursor-pointer"
        >
          <input
            class="cursor-pointer"
            type="checkbox"
            value="work"
          />
          work
        </label>
      </div>
      <div
        class="mt-4 p-3 rounded-lg "
      >
        <strong>
          Seleccionados:
        </strong>
         
        Ninguno
      </div>
      <label
        for="file-upload"
      >
        Sube una foto
      </label>
      <input
        class="bg-sky-900 p-2.5 w-full sm:w-100 cursor-pointer "
        id="file-upload"
        type="file"
      />
      <button
        class="sc-FEMpB dyFOCA mt-4 mb-2.5"
        type="submit"
      >
        Publicar
      </button>
    </form>
  </div>
</div>
`;
